core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/fuse/PlanNodeFuser.java

```diff
@@ -0,0 +1,530 @@
+    {
+        if (optionalMaskSymbol.isEmpty()) {
+            return otherMaskSymbol.isEmpty() && otherMask == null;
+        }
+        Symbol maskSymbol = optionalMaskSymbol.get();
+        Expression mask = masks.get(maskSymbol);
+        if (mask == null) {
+            // just not fused mask. symbols must match
+            return maskSymbol.equals(otherMaskSymbol.orElse(null)) && otherMask == null;
+        }
+        return mask.equals(otherMask);
+    }
+
+    /**
+     * Fuses two table scan nodes.
+     * Fusing two table scans succeeds only if both scan the same table (T).
+     * The fused result is defined as:
+     * (Scan(T), columnMap(right, left), TRUE, TRUE)
+     * where columnMap(right, left) maps right symbol to the existing left symbol if it exists.
+     *
+     * <p>
+     * As an example, fusing two query fragments that scan the same table:
+     * <pre>
+     * SELECT i_item_sk AS sk, i_brand AS brand
+     * FROM item
+     *
+     * SELECT i_brand AS brand2, i_size AS size
+     * FROM item
+     * </pre>
+     * results in a fused result (P, M, TRUE, TRUE) where:
+     * <pre>
+     * P: SELECT
+     *   i_item_sk AS sk,
+     *   i_brand AS brand,
+     *   i_size AS size
+     * FROM item
+     * M: brand2 → brand
+     * </pre>
+     * </p>
+     */
+    private Optional<FusedPlanNode> fuseTableScan(TableScanNode left, TableScanNode right)
+    {
+        if (!left.getTable().equals(right.getTable())) {
+            return Optional.empty();
+        }
+
+        if (!left.getEnforcedConstraint().equals(right.getEnforcedConstraint()) ||
+                !left.getStatistics().equals(right.getStatistics()) ||
+                !left.isUpdateTarget() == right.isUpdateTarget() ||
+                !left.getUseConnectorNodePartitioning().equals(right.getUseConnectorNodePartitioning())) {
+            return Optional.empty();
+        }
+
+        Set<Symbol> outputs = new HashSet<>(left.getOutputSymbols());
+        Map<Symbol, ColumnHandle> assignments = new HashMap<>(left.getAssignments());
+        FusedSymbolMapping.Builder symbolMappingBuilder = FusedSymbolMapping.builder();
+        Set<Symbol> rightOutputSymbols = ImmutableSet.copyOf(right.getOutputSymbols());
+
+        // add unique assignments and outputs from the right side or map to the left side symbols
+        right.getAssignments().forEach(((symbol, columnHandle) ->
+                left.getAssignments().entrySet().stream().filter(entry -> entry.getValue().equals(columnHandle)).findFirst()
+                        .ifPresentOrElse(
+                                entry -> {
+                                    // left assignment found, add the mapping and output symbol if necessary
+                                    symbolMappingBuilder.add(symbol, entry.getKey());
+                                    if (rightOutputSymbols.contains(symbol)) {
```

Comment:
Isn't this always true since we require that `assignments.keySet().containsAll(outputs)` in table scan node ?

Reply:
I'm thinking why don't we loop on `rightOutputSymbols` instead and only lookup the assignments which are part of `rightOutputSymbols` ? I don't think we need to care about the assignments which are not part of output symbols as `PruneTableScanColumns` would've gotten rid of them anyway.

Reply:
ping

Reply:
why does `TableScanNode` have separate output symbols if they are not needed?
I feel like prunning the assignments here is not the right place because this rule shouldn't care if assignments are needed or not.